File: com.chartboost.helium/Android~/helium-android-bridge/src/main/java/com/chartboost/heliumsdk/unity/HeliumUnityBridge.java
package com.chartboost.heliumsdk.unity;

import static com.chartboost.heliumsdk.unity.HeliumEventProcessor.serializeHeliumBidEvent;
import static com.chartboost.heliumsdk.unity.HeliumEventProcessor.serializeHeliumEvent;
import static com.chartboost.heliumsdk.unity.HeliumEventProcessor.serializeHeliumRewardEvent;
import static com.chartboost.heliumsdk.unity.HeliumEventProcessor.serializePlacementILRDData;
import static com.chartboost.heliumsdk.unity.HeliumUnityAdWrapper.Wrap;
import android.app.Activity;
import android.util.Log;
import com.chartboost.heliumsdk.HeliumSdk;
import com.chartboost.heliumsdk.HeliumIlrdObserver;
import com.chartboost.heliumsdk.ad.HeliumAdError;
import com.chartboost.heliumsdk.ad.HeliumBannerAd;
import com.chartboost.heliumsdk.ad.HeliumBannerAdListener;
import com.chartboost.heliumsdk.ad.HeliumInterstitialAd;
import com.chartboost.heliumsdk.ad.HeliumInterstitialAdListener;
import com.chartboost.heliumsdk.ad.HeliumRewardedAd;
import com.chartboost.heliumsdk.ad.HeliumRewardedAdListener;
import com.unity3d.player.UnityPlayer;
import androidx.annotation.NonNull;
import java.util.HashMap;

public class HeliumUnityBridge {
    private static final String TAG = "HeliumUnityBridge";
    private static final String EMPTY_STRING = "";

    private Activity _activity;

    private ILifeCycleEventListener lifeCycleEventListener;
    private IBannerEventListener bannerEventsListener;
    private IInterstitialEventListener interstitialEventsListener;
    private IRewardedEventListener rewardedEventListener;
    private HeliumIlrdObserver ilrdObserver;

    // Stores a static instance of the HeliumPlugin class for easy access
    // from Unity
    @SuppressWarnings("unused")
    public static HeliumUnityBridge instance() {
        return new HeliumUnityBridge();
    }

    @SuppressWarnings("unused")
    public void setupEventListeners(final ILifeCycleEventListener lifeCycleListener,
                                    final IInterstitialEventListener interstitialListener,
                                    final IRewardedEventListener rewardedListener,
                                    final IBannerEventListener bannerListener)
    {
        lifeCycleEventListener = lifeCycleListener;
        interstitialEventsListener = interstitialListener;
        rewardedEventListener = rewardedListener;
        bannerEventsListener = bannerListener;
    }

    @SuppressWarnings("unused")
    public void setSubjectToCoppa(boolean isSubject) {
        HeliumSdk.setSubjectToCoppa(isSubject);
    }

    @SuppressWarnings("unused")
    public void setSubjectToGDPR(boolean isSubject) {
        HeliumSdk.setSubjectToGDPR(isSubject);
    }

    @SuppressWarnings("unused")
    public void setCCPAConsent(boolean hasGivenConsent) { HeliumSdk.setCCPAConsent(hasGivenConsent); }

    @SuppressWarnings("unused")
    public void setUserHasGivenConsent(boolean hasGivenConsent) { HeliumSdk.setUserHasGivenConsent(hasGivenConsent); }

    @SuppressWarnings("unused")
    public void setUserIdentifier(final String userIdentifier) { HeliumSdk.setUserIdentifier(userIdentifier); }

    @SuppressWarnings("unused")
    public String getUserIdentifier() {
        return HeliumSdk.getUserIdentifier();
    }

    @SuppressWarnings("unused")
    public boolean onBackPressed() {
       return HeliumSdk.onBackPressed();
    }

    @SuppressWarnings("unused")
    public void start(final String appId, final String appSignature, final String unityVersion) {
        _activity = UnityPlayer.currentActivity;
        ilrdObserver = impData -> {
            String json = serializePlacementILRDData(impData.getPlacementId(), impData.getIlrdInfo());
            lifeCycleEventListener.DidReceiveILRD(json);
        };
        runTaskOnUiThread(() -> {
            // This call initializes the Helium SDK. This might change in the future with two ID parameters and we'll get rid of the logControllerListener
            HeliumSdk.start(_activity, appId, appSignature, error -> {
                boolean errorNotFound = error == null;
                if (errorNotFound) {
                    Log.d("Unity", "HeliumUnityBridge: Plugin Initialized");
                    HeliumSdk.setGameEngine("unity", unityVersion);
                    HeliumSdk.subscribeIlrd(ilrdObserver);
                } else {
                    Log.d("Unity", "HeliumUnityBridge: Plugin failed to initialize: " + error);
                }

                int errorCode = errorNotFound ? -1 : 1;
                String errorDescription = errorNotFound ? EMPTY_STRING : error.toString();
                lifeCycleEventListener.DidStart(errorCode, errorDescription);
            });
        });
    }

    @SuppressWarnings("unused")
    public void setTestMode(boolean mode) {
        HeliumSdk.setTestMode(mode);
    }

    @SuppressWarnings("unused")
    public void destroy() {
        if (ilrdObserver != null) {
            HeliumSdk.unsubscribeIlrd(ilrdObserver);
            ilrdObserver = null;
        }
    }

    @SuppressWarnings("unused")
    public HeliumUnityAdWrapper getInterstitialAd(final String placementName) {
        if (placementName == null)
            return null;

        return Wrap(new HeliumInterstitialAd(placementName, new HeliumInterstitialAdListener() {
            @Override
            public void didReceiveWinningBid(@NonNull String placementName, @NonNull HashMap<String, String> hashMap) {
                serializeHeliumBidEvent(placementName, hashMap, interstitialEventsListener::DidWinBidInterstitial);
            }

            @Override
            public void didCache(@NonNull String placementName, HeliumAdError error) {
                serializeHeliumEvent(placementName,  error, interstitialEventsListener::DidLoadInterstitial);
            }

            @Override
            public void didShow(@NonNull String placementName, HeliumAdError error) {
                serializeHeliumEvent(placementName,  error, interstitialEventsListener::DidShowInterstitial);
            }

            @Override
            public void didClose(@NonNull String placementName, HeliumAdError error) {
                serializeHeliumEvent(placementName,  error, interstitialEventsListener::DidCloseInterstitial);
            }

            @Override
            public void didClick(@NonNull String placementName, HeliumAdError error) {
                serializeHeliumEvent(placementName,  error, interstitialEventsListener::DidClickInterstitial);
            }

            @Override
            public void didRecordImpression(@NonNull String placementName) {
                serializeHeliumEvent(placementName, null, interstitialEventsListener::DidRecordImpression);
            }
        }));
    }

    @SuppressWarnings("unused")
    public HeliumUnityAdWrapper getRewardedAd(final String placementName) {
        if (placementName == null)
            return null;

        return Wrap(new HeliumRewardedAd(placementName, new HeliumRewardedAdListener() {
            @Override
            public void didReceiveWinningBid(@NonNull String placementName, @NonNull HashMap<String, String> hashMap) {
                serializeHeliumBidEvent(placementName, hashMap, rewardedEventListener::DidWinBidRewarded);
            }

            @Override
            public void didCache(@NonNull String placementName, HeliumAdError error) {
                serializeHeliumEvent(placementName,  error, rewardedEventListener::DidLoadRewarded);
            }

            @Override
            public void didShow(@NonNull String placementName, HeliumAdError error) {
                serializeHeliumEvent(placementName,  error, rewardedEventListener::DidShowRewarded);
            }

            @Override
            public void didClose(@NonNull String placementName, HeliumAdError error) {
                serializeHeliumEvent(placementName,  error, rewardedEventListener::DidCloseRewarded);
            }

            @Override
            public void didReceiveReward(@NonNull String placementName, @NonNull String reward) {
                serializeHeliumRewardEvent(placementName,  reward, rewardedEventListener::DidReceiveReward);
            }

            @Override
            public void didClick(@NonNull String placementName, HeliumAdError error) {
                serializeHeliumEvent(placementName,  error, rewardedEventListener::DidClickRewarded);
            }

            @Override
            public void didRecordImpression(@NonNull String placementName){
                serializeHeliumEvent(placementName, null, rewardedEventListener::DidRecordImpression);
            }
        }));
    }

    @SuppressWarnings("unused")
    public HeliumUnityAdWrapper getBannerAd(final String placementName, final int size) {
        if (placementName == null)
            return null;

        HeliumBannerAd.HeliumBannerSize wantedSize = null;

        switch (size) {
            case 0:
                wantedSize = HeliumBannerAd.HeliumBannerSize.STANDARD;
                break;
            case 1:
                wantedSize = HeliumBannerAd.HeliumBannerSize.MEDIUM;
                break;
            case 2:
                wantedSize = HeliumBannerAd.HeliumBannerSize.LEADERBOARD;
                break;
        }
        HeliumBannerAd.HeliumBannerSize finalWantedSize = wantedSize;

        return Wrap(new HeliumBannerAd(_activity, placementName, finalWantedSize, new HeliumBannerAdListener() {
            @Override
            public void didReceiveWinningBid(@NonNull String placementName, @NonNull HashMap<String, String> hashMap) {
                serializeHeliumBidEvent(placementName, hashMap, bannerEventsListener::DidWinBidBanner);
            }

            @Override
            public void didCache(@NonNull String placementName, HeliumAdError error) {
                serializeHeliumEvent(placementName,  error, bannerEventsListener::DidLoadBanner);
            }

            @Override
            public void didClick(@NonNull String placementName, HeliumAdError error) {
                serializeHeliumEvent(placementName,  error, bannerEventsListener::DidClickBanner);
            }

            @Override
            public void didRecordImpression(@NonNull String placementName) {
                serializeHeliumEvent(placementName, null, bannerEventsListener::DidRecordImpression);
            }
        }));
    }

    // Method whenever work needs to be done on the UI thread.
    public static void runTaskOnUiThread(final Runnable runnable) {
        UnityPlayer.currentActivity.runOnUiThread(() -> {
            try {
                runnable.run();
            } catch (Exception ex) {
                Log.w(TAG, "Exception found when running on UI Thread" + ex.getMessage());
            }
        });
    }
}

File: com.chartboost.helium/Android~/helium-android-bridge/src/main/kotlin/com/chartboost/heliumsdk/unity/HeliumUnityBridge.kt
package com.chartboost.heliumsdk.unity

import android.app.Activity
import android.util.Log
import com.chartboost.heliumsdk.HeliumIlrdObserver
import com.chartboost.heliumsdk.HeliumImpressionData
import com.chartboost.heliumsdk.HeliumSdk
import com.chartboost.heliumsdk.ad.*
import com.chartboost.heliumsdk.ad.HeliumBannerAd.HeliumBannerSize
import com.chartboost.heliumsdk.unity.HeliumEventProcessor.HeliumBidEventConsumer
import com.chartboost.heliumsdk.unity.HeliumEventProcessor.HeliumEventConsumer
import com.chartboost.heliumsdk.unity.HeliumEventProcessor.HeliumRewardEventConsumer
import com.chartboost.heliumsdk.unity.HeliumEventProcessor.serializeHeliumBidEvent
import com.chartboost.heliumsdk.unity.HeliumEventProcessor.serializeHeliumEvent
import com.chartboost.heliumsdk.unity.HeliumEventProcessor.serializeHeliumRewardEvent
import com.chartboost.heliumsdk.unity.HeliumEventProcessor.serializePlacementILRDData
import com.chartboost.heliumsdk.unity.HeliumUnityAdWrapper.Companion.wrap
import com.unity3d.player.UnityPlayer

@Suppress("NAME_SHADOWING")
class HeliumUnityBridge {
    private lateinit var _activity: Activity
    private var lifeCycleEventListener: ILifeCycleEventListener? = null
    private var bannerEventsListener: IBannerEventListener? = null
    private var interstitialEventsListener: IInterstitialEventListener? = null
    private var rewardedEventListener: IRewardedEventListener? = null
    private var ilrdObserver: HeliumIlrdObserver? = null
    fun setupEventListeners(
        lifeCycleListener: ILifeCycleEventListener?,
        interstitialListener: IInterstitialEventListener?,
        rewardedListener: IRewardedEventListener?,
        bannerListener: IBannerEventListener?
    ) {
        lifeCycleEventListener = lifeCycleListener
        interstitialEventsListener = interstitialListener
        rewardedEventListener = rewardedListener
        bannerEventsListener = bannerListener
    }

    fun setSubjectToCoppa(isSubject: Boolean) {
        HeliumSdk.setSubjectToCoppa(isSubject)
    }

    fun setSubjectToGDPR(isSubject: Boolean) {
        HeliumSdk.setSubjectToGDPR(isSubject)
    }

    fun setCCPAConsent(hasGivenConsent: Boolean) {
        HeliumSdk.setCCPAConsent(hasGivenConsent)
    }

    fun setUserHasGivenConsent(hasGivenConsent: Boolean) {
        HeliumSdk.setUserHasGivenConsent(hasGivenConsent)
    }

    var userIdentifier: String?
        get() = HeliumSdk.getUserIdentifier()
        set(userIdentifier) {
            HeliumSdk.setUserIdentifier(userIdentifier)
        }

    fun onBackPressed(): Boolean {
        return HeliumSdk.onBackPressed()
    }

    fun start(appId: String?, appSignature: String?, unityVersion: String?) {
        _activity = UnityPlayer.currentActivity

        ilrdObserver = object : HeliumIlrdObserver {
            override fun onImpression(impData: HeliumImpressionData) {
                val json = serializePlacementILRDData(impData.placementId, impData.ilrdInfo)
                lifeCycleEventListener!!.DidReceiveILRD(json)
            }
        }
        runTaskOnUiThread {
            // This call initializes the Helium SDK. This might change in the future with two ID parameters and we'll get rid of the logControllerListener
            HeliumSdk.start(_activity, appId!!, appSignature!!) { error: Error? ->
                val errorNotFound = error == null
                if (errorNotFound) {
                    Log.d("Unity", "HeliumUnityBridge: Plugin Initialized")
                    HeliumSdk.setGameEngine("unity", unityVersion)
                    HeliumSdk.subscribeIlrd(ilrdObserver!!)
                } else {
                    Log.d("Unity", "HeliumUnityBridge: Plugin failed to initialize: $error")
                }
                val errorCode = if (errorNotFound) -1 else 1
                val errorDescription = if (errorNotFound) EMPTY_STRING else error.toString()
                lifeCycleEventListener!!.DidStart(errorCode, errorDescription)
            }
        }
    }

    fun setTestMode(mode: Boolean) {
        HeliumSdk.setTestMode(mode)
    }

    fun destroy() {
        if (ilrdObserver != null) {
            HeliumSdk.unsubscribeIlrd(ilrdObserver!!)
            ilrdObserver = null
        }
    }

    fun getInterstitialAd(placementName: String?): HeliumUnityAdWrapper? {
        return if (placementName == null) null else wrap(
            HeliumInterstitialAd(
                placementName,
                object : HeliumInterstitialAdListener {
                    override fun didReceiveWinningBid(
                        placementName: String,
                        hashMap: HashMap<String, String>
                    ) {
                        serializeHeliumBidEvent(
                            placementName,
                            hashMap,
                            HeliumBidEventConsumer { placementName: String?, auctionId: String?, partnerId: String?, price: Double ->
                                interstitialEventsListener!!.DidWinBidInterstitial(
                                    placementName,
                                    auctionId,
                                    partnerId,
                                    price
                                )
                            })
                    }

                    override fun didCache(placementName: String, error: HeliumAdError?) {
                        serializeHeliumEvent(
                            placementName,
                            error,
                            HeliumEventConsumer { placementName: String?, errorCode: Int, errorDescription: String? ->
                                interstitialEventsListener!!.DidLoadInterstitial(
                                    placementName,
                                    errorCode,
                                    errorDescription
                                )
                            })
                    }

                    override fun didShow(placementName: String, error: HeliumAdError?) {
                        serializeHeliumEvent(
                            placementName,
                            error,
                            HeliumEventConsumer { placementName: String?, errorCode: Int, errorDescription: String? ->
                                interstitialEventsListener!!.DidShowInterstitial(
                                    placementName,
                                    errorCode,
                                    errorDescription
                                )
                            })
                    }

                    override fun didClose(placementName: String, error: HeliumAdError?) {
                        serializeHeliumEvent(
                            placementName,
                            error,
                            HeliumEventConsumer { placementName: String?, errorCode: Int, errorDescription: String? ->
                                interstitialEventsListener!!.DidCloseInterstitial(
                                    placementName,
                                    errorCode,
                                    errorDescription
                                )
                            })
                    }

                    override fun didClick(placementName: String, error: HeliumAdError?) {
                        serializeHeliumEvent(
                            placementName,
                            error,
                            HeliumEventConsumer { placementName: String?, errorCode: Int, errorDescription: String? ->
                                interstitialEventsListener!!.DidClickInterstitial(
                                    placementName,
                                    errorCode,
                                    errorDescription
                                )
                            })
                    }

                    override fun didRecordImpression(placementName: String) {
                        serializeHeliumEvent(
                            placementName,
                            null,
                            HeliumEventConsumer { placementName: String?, errorCode: Int, errorDescription: String? ->
                                interstitialEventsListener!!.DidRecordImpression(
                                    placementName,
                                    errorCode,
                                    errorDescription
                                )
                            })
                    }
                })
        )
    }

    fun getRewardedAd(placementName: String?): HeliumUnityAdWrapper? {
        return if (placementName == null) null else wrap(
            HeliumRewardedAd(
                placementName,
                object : HeliumRewardedAdListener {
                    override fun didReceiveWinningBid(
                        placementName: String,
                        hashMap: HashMap<String, String>
                    ) {
                        serializeHeliumBidEvent(
                            placementName,
                            hashMap,
                            HeliumBidEventConsumer { placementName: String?, auctionId: String?, partnerId: String?, price: Double ->
                                rewardedEventListener!!.DidWinBidRewarded(
                                    placementName,
                                    auctionId,
                                    partnerId,
                                    price
                                )
                            })
                    }

                    override fun didCache(placementName: String, error: HeliumAdError?) {
                        serializeHeliumEvent(
                            placementName,
                            error,
                            HeliumEventConsumer { placementName: String?, errorCode: Int, errorDescription: String? ->
                                rewardedEventListener!!.DidLoadRewarded(
                                    placementName,
                                    errorCode,
                                    errorDescription
                                )
                            })
                    }

                    override fun didShow(placementName: String, error: HeliumAdError?) {
                        serializeHeliumEvent(
                            placementName,
                            error,
                            HeliumEventConsumer { placementName: String?, errorCode: Int, errorDescription: String? ->
                                rewardedEventListener!!.DidShowRewarded(
                                    placementName,
                                    errorCode,
                                    errorDescription
                                )
                            })
                    }

                    override fun didClose(placementName: String, error: HeliumAdError?) {
                        serializeHeliumEvent(
                            placementName,
                            error,
                            HeliumEventConsumer { placementName: String?, errorCode: Int, errorDescription: String? ->
                                rewardedEventListener!!.DidCloseRewarded(
                                    placementName,
                                    errorCode,
                                    errorDescription
                                )
                            })
                    }

                    override fun didReceiveReward(placementName: String, reward: String) {
                        serializeHeliumRewardEvent(
                            placementName,
                            reward,
                            HeliumRewardEventConsumer { placementName: String?, reward: Int ->
                                rewardedEventListener!!.DidReceiveReward(
                                    placementName,
                                    reward
                                )
                            })
                    }

                    override fun didClick(placementName: String, error: HeliumAdError?) {
                        serializeHeliumEvent(
                            placementName,
                            error,
                            HeliumEventConsumer { placementName: String?, errorCode: Int, errorDescription: String? ->
                                rewardedEventListener!!.DidClickRewarded(
                                    placementName,
                                    errorCode,
                                    errorDescription
                                )
                            })
                    }

                    override fun didRecordImpression(placementName: String) {
                        serializeHeliumEvent(
                            placementName,
                            null,
                            HeliumEventConsumer { placementName: String?, errorCode: Int, errorDescription: String? ->
                                rewardedEventListener!!.DidRecordImpression(
                                    placementName,
                                    errorCode,
                                    errorDescription
                                )
                            })
                    }
                })
        )
    }

    fun getBannerAd(placementName: String?, size: Int): HeliumUnityAdWrapper? {
        if (placementName == null) return null
        var wantedSize: HeliumBannerSize? = null
        when (size) {
            0 -> wantedSize = HeliumBannerSize.STANDARD
            1 -> wantedSize = HeliumBannerSize.MEDIUM
            2 -> wantedSize = HeliumBannerSize.LEADERBOARD
        }
        val finalWantedSize = wantedSize
        return wrap(
            HeliumBannerAd(
                _activity,
                placementName,
                finalWantedSize!!,
                object : HeliumBannerAdListener {
                    override fun didReceiveWinningBid(
                        placementName: String,
                        bidInfo: HashMap<String, String>
                    ) {
                        serializeHeliumBidEvent(placementName, bidInfo,
                            HeliumBidEventConsumer { placementName: String?, auctionId: String?, partnerId: String?, price: Double ->
                                bannerEventsListener!!.DidWinBidBanner(
                                    placementName,
                                    auctionId,
                                    partnerId,
                                    price
                                )
                            })
                    }

                    override fun didCache(placementName: String, error: HeliumAdError?) {
                        serializeHeliumEvent(
                            placementName,
                            error,
                            HeliumEventConsumer { placementName: String?, errorCode: Int, errorDescription: String? ->
                                bannerEventsListener!!.DidLoadBanner(
                                    placementName,
                                    errorCode,
                                    errorDescription
                                )
                            })
                    }

                    override fun didClick(placementName: String, error: HeliumAdError?) {
                        serializeHeliumEvent(
                            placementName,
                            error,
                            HeliumEventConsumer { placementName: String?, errorCode: Int, errorDescription: String? ->
                                bannerEventsListener!!.DidClickBanner(
                                    placementName,
                                    errorCode,
                                    errorDescription
                                )
                            })
                    }

                    override fun didRecordImpression(placementName: String) {
                        serializeHeliumEvent(
                            placementName,
                            null,
                            HeliumEventConsumer { placementName: String?, errorCode: Int, errorDescription: String? ->
                                bannerEventsListener!!.DidRecordImpression(
                                    placementName,
                                    errorCode,
                                    errorDescription
                                )
                            })
                    }
                })
        )
    }

    companion object {
        private const val TAG = "HeliumUnityBridge"
        private const val EMPTY_STRING = ""

        // Stores a static instance of the HeliumPlugin class for easy access
        // from Unity
        fun instance(): HeliumUnityBridge {
            return HeliumUnityBridge()
        }

        // Method whenever work needs to be done on the UI thread.
        fun runTaskOnUiThread(runnable: Runnable) {
            UnityPlayer.currentActivity.runOnUiThread {
                try {
                    runnable.run()
                } catch (ex: Exception) {
                    Log.w(TAG, "Exception found when running on UI Thread" + ex.message)
                }
            }
        }
    }
}
